### PR TITLE
fuchsia: Support multiple formats of Vulkan surface images.

### PR DESCRIPTION
**Pull request by gnoliyil**
Change https://github.com/flutter/engine/pull/23488/
has switched all the Vulkan surfaces on Fuchsia to use
RGBA8 color type. However, Using RGBA8 format on Intel GPUs
has caused graphical artifacts on emulators
(See http://fxbug.dev/70232).

This change modifies vulkan_surface.cc to use the multi-format
vkSetBufferCollectionImageConstraintsFUCHSIA() Vulkan API,
so that it could specify multiple formats when allocating
buffer collections.

Bug: fxbug.dev/70232

### CODE REVIEW

**Comment by gnoliyil** [comment]
dragostis@ dreveman@ Could you please take a look?

**Review by dragostis — APPROVED** [review]

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 1]
```diff
@@ -26,9 +26,21 @@ namespace flutter_runner {
...
+// until we have a more reliable API for creating external Vulkan
+// images using sysmem. TODO(fxb/52835)
+#if defined(__aarch64__)
 constexpr SkColorType kSkiaColorType = kRGBA_8888_SkColorType;
 constexpr VkFormat kVulkanFormat = VK_FORMAT_R8G8B8A8_UNORM;
 constexpr VkImageCreateFlags kVulkanImageCreateFlags = 0;
+#else
+constexpr SkColorType kSkiaColorType = kBGRA_8888_SkColorType;
+constexpr VkFormat kVulkanFormat = VK_FORMAT_B8G8R8A8_UNORM;
+constexpr VkImageCreateFlags kVulkanImageCreateFlags =
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:40]: why is this needed? it was only needed prior to sysmem because we had to match the parameters used by scenic for import

**Review by dreveman — CHANGES_REQUESTED** [review]
If we need to support BGRA then please use setBufferCollectionImageConstraintsFUCHSIA instead and set multiple formats. That way we can allow the driver to decide what format is ideal and keep this type of workaround (if needed) in a more appropriate location.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc (hunk 1 above)
Comment (gnoliyil) [reply shell/platform/fuchsia/flutter/vulkan_surface.cc]: Discussed offline. We decided to use multi-format sysmem extensions so that drivers could choose between BGRA and RGBA formats.

**Comment by gnoliyil** [comment]
> If we need to support BGRA then please use setBufferCollectionImageConstraintsFUCHSIA instead and set multiple formats. That way we can allow the driver to decide what format is ideal and keep this type of workaround (if needed) in a more appropriate location.

Done. PTAL.

**Comment by gnoliyil** [comment]
This change won't build until https://github.com/flutter/engine/pull/24829 and https://github.com/flutter/buildroot/pull/441 are submitted to support the latest Vulkan headers in Flutter repository.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 2]
```diff
@@ -26,9 +26,19 @@ namespace flutter_runner {
...
+// 
+// #if defined(__aarch64__)
+// constexpr SkColorType kSkiaColorType = kRGBA_8888_SkColorType;
+// constexpr VkFormat kVulkanFormat = VK_FORMAT_R8G8B8A8_UNORM;
+// constexpr VkImageCreateFlags kVulkanImageCreateFlags = 0;
+// #else
+// constexpr VkImageCreateFlags kVulkanImageCreateFlags =
+//     VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT;
+// #endif
+
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:41]: please remove this before landing

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 3]
```diff
@@ -50,33 +60,87 @@ bool VulkanSurface::CreateVulkanImage(vulkan::VulkanProvider& vulkan_provider,
...
-      .mipLevels = 1,
-      .arrayLayers = 1,
-      .samples = VK_SAMPLE_COUNT_1_BIT,
-      .tiling = VK_IMAGE_TILING_OPTIMAL,
-      .usage = kVkImageUsage,
-      .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
-      .queueFamilyIndexCount = 0,
-      .pQueueFamilyIndices = nullptr,
-      .initialLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+  const VkImageCreateInfo kDefaultImageCreateInfo = {
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:80]: this is not really a global constant as you're setting pnext and extent to local values. please rename to image_create_info or maybe better, fold it into the loop I'm suggesting below

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 4]
```diff
@@ -50,33 +60,87 @@ bool VulkanSurface::CreateVulkanImage(vulkan::VulkanProvider& vulkan_provider,
...
+    .samples = VK_SAMPLE_COUNT_1_BIT,
+    .tiling = VK_IMAGE_TILING_OPTIMAL,
+    .usage = kVkImageUsage,
+    .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
+    .queueFamilyIndexCount = 0,
+    .pQueueFamilyIndices = nullptr,
+    .initialLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+  };
+
+  constexpr size_t kNumFormats = 2u;
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:99]: please replace this with an array of VK_FORMATs and iterate over it to build create_infos.

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 5]
```diff
@@ -50,33 +60,87 @@ bool VulkanSurface::CreateVulkanImage(vulkan::VulkanProvider& vulkan_provider,
...
+  std::vector<VkImageFormatConstraintsInfoFUCHSIA> format_constraints(kNumFormats, kDefaultFormatInfo);
+  std::vector<VkImageCreateInfo> create_infos(kNumFormats, kDefaultImageCreateInfo);
+  // Immutable format is technically limited to R8G8B8A8_SRGB but
+  // R8G8B8A8_UNORM works with existing ARM drivers so we allow that
+  // until we have a more reliable API for creating external Vulkan
+  // images using sysmem. TODO(fxb/52835)
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:105]: this commend doesn't make sense when using sysmem. why are you adding this back?

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 6]
```diff
@@ -50,33 +60,87 @@ bool VulkanSurface::CreateVulkanImage(vulkan::VulkanProvider& vulkan_provider,
...
+  format_constraints[0].sysmemFormat = static_cast<uint64_t>(fuchsia::sysmem::PixelFormatType::R8G8B8A8);
+  create_infos[0].format = VK_FORMAT_R8G8B8A8_UNORM;
+  format_constraints[1].sysmemFormat = static_cast<uint64_t>(fuchsia::sysmem::PixelFormatType::BGRA32);
+  create_infos[1].format = VK_FORMAT_B8G8R8A8_UNORM;
+  create_infos[1].flags = VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT;
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:110]: why add this flag? it shouldn't be needed when using sysmem

File: shell/platform/fuchsia/flutter/vulkan_surface.cc [hunk 7]
```diff
@@ -50,33 +60,87 @@ bool VulkanSurface::CreateVulkanImage(vulkan::VulkanProvider& vulkan_provider,
...
+
+  const VkImageConstraintsInfoFUCHSIA constraints_info = {
+    .sType = VK_STRUCTURE_TYPE_IMAGE_CONSTRAINTS_INFO_FUCHSIA,
+    .pNext = nullptr,
+    .createInfoCount = kNumFormats,
```
Comment (dreveman) [inline shell/platform/fuchsia/flutter/vulkan_surface.cc:115]: nit: create_infos.size()

**Review by dreveman — CHANGES_REQUESTED** [review]

**Comment by chinmaygarde** [comment]
@gnoliyil Any updates on @dreveman s last review?

**Comment by dreveman** [comment]
We found a better way to solve this. In the driver. So this is no longer needed.
